src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs

```diff
@@ -76,65 +75,63 @@ public bool ExecuteCommand(GoToDefinitionCommandArgs args, CommandExecutionConte
-                return await asyncService.TryGoToDefinitionAsync(document, position, cancellationToken).ConfigureAwait(false);
+                // We're showing our own UI, ensure the editor doesn't show anything itself.
+                context.OperationContext.TakeOwnership();
+                var token = _listener.BeginAsyncOperation(nameof(ExecuteCommand));
+                ExecuteModernCommandAsync(args, document, asyncService, caretPos.Value).CompletesAsyncOperation(token);
             }
             else
             {
-                succeeded = await ExecuteCommandLegacyAsync();
+                bool succeeded;
+                using (context.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Navigating_to_definition))
+                {
+                    succeeded = service.TryGoToDefinition(document, caretPos.Value, context.OperationContext.UserCancellationToken);
+                }
+
+                if (!succeeded)
+                    ReportFailure(document);
             }
 
+            return true;
+        }
+
+        private static void ReportFailure(Document document)
+        {
+            var notificationService = document.Project.Solution.Workspace.Services.GetRequiredService<INotificationService>();
+            notificationService.SendNotification(
+                FeaturesResources.Cannot_navigate_to_the_symbol_under_the_caret, EditorFeaturesResources.Go_to_Definition, NotificationSeverity.Information);
+        }
+
+        private async Task ExecuteModernCommandAsync(
+            GoToDefinitionCommandArgs args, Document document, IAsyncGoToDefinitionService service, SnapshotPoint position)
+        {
+            bool succeeded;
+
+            var indicatorFactory = document.Project.Solution.Workspace.Services.GetRequiredService<IBackgroundWorkIndicatorFactory>();
             using (var backgroundIndicator = indicatorFactory.Create(
                 args.TextView, new SnapshotSpan(args.SubjectBuffer.CurrentSnapshot, position, 1),
                 EditorFeaturesResources.Navigating_to_definition))
             {
                 await Task.Delay(5000).ConfigureAwait(false);
-                succeeded = await GoToDefinitionAsync(
-                    document, position, service, backgroundIndicator.UserCancellationToken).ConfigureAwait(false);
+
+                // determine the location first.
+                var cancellationToken = backgroundIndicator.UserCancellationToken;
+                var location = await service.GetDefinitionLocationAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+                // we're about to navigate.  so disable navigation cancellation in our indicator so we don't self-cancel.
+                backgroundIndicator.CancelOnNavigation = false;
+                succeeded = location != null && await location.TryNavigateToAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
It would be nice if `location.TryNavigateToAsync` had an overload that took the `IBackgroundWorkIndicatorContext` and did the right thing.